src/red_black_tree.rs

```diff
@@ -331,235 +443,257 @@ impl<
+        let left = self.get_left(node_index);
+        let right = self.get_right(node_index);
+        let (pivot_node_index, parent_and_dir) = if self.is_leaf(node_index) {
+            if !self.is_root(node_index) {
+                let parent = self.get_parent(node_index);
+                let dir = self._child_dir(parent, node_index);
+                // Remove pointer to the removed leaf node
+                self._connect(parent, SENTINEL, dir);
+                (SENTINEL, Some((parent, dir)))
+            } else {
+                // Set the root to SENTINEL
+                self.root = SENTINEL;
+                (SENTINEL, None)
+            }
+        } else if left == SENTINEL {
+            self._transplant(node_index, right);
+            (right, None)
+        } else if right == SENTINEL {
+            self._transplant(node_index, left);
+            (left, None)
+        } else {
+            // Find the largest node in the left subtree
+            let mut parent_and_dir = None;
+            let max_left = self._find_max(left);
+            let max_left_parent = self.get_parent(max_left);
+            let max_left_child = self.get_left(max_left);
+            is_black = self.is_black(max_left);
+
+            // If max_left is not equal to root of the left subtree, then
+            // replace the root of the left subtree with max_left and replace
+            // max_left with max_left_child
+            if self.get_parent(max_left) != node_index {
+                self._transplant(max_left, max_left_child);
+                // We perform this operation in the conditional because we do not
+                // want to form a cycle
+                self._connect(max_left, self.get_left(node_index), Field::Left as u32);
+                if max_left_child == SENTINEL {
+                    parent_and_dir = Some((max_left_parent, Field::Right as u32));
+                }
+            } else if max_left_child == SENTINEL {
+                // The only time this is called is is when the left subtree is
+                // a single node
+                assert!(self.is_leaf(max_left));
+                parent_and_dir = Some((max_left, Field::Left as u32));
+            }
+
+            // Complete the transplant of max_left
+            self._transplant(node_index, max_left);
+            self._connect(max_left, self.get_right(node_index), Field::Right as u32);
+
+            self._color_node(max_left, self.get_color(node_index));
+
+            (max_left_child, parent_and_dir)
+        };
+
+        // Completely remove the current node index from the tree
+        self._remove_allocator_node(node_index);
+
+        if is_black && !self.is_root(pivot_node_index) {
+            self._fix_remove(pivot_node_index, parent_and_dir);
         }
     }
 
+    fn _fix_remove(&mut self, mut node_index: u32, parent_and_dir: Option<(u32, u32)>) {
+        let (mut parent, mut dir) = parent_and_dir.unwrap_or({
```

Comment:
Issue (2a): if the `node_index` is SENTINEL, we still need to rotate the tree to rebalance. This problem is solved by passing in a optional tuple with the hypothetical "parent" of the SENTINEL node and the direction